File: ycs-golang/ycs_bootstrap.go
package main

import (
	"sync"
	"ycs/content"
	"ycs/contracts"
	"ycs/core"
)

// YcsBootstrap provides bootstrap functionality to initialize dependencies and break circular references
type YcsBootstrap struct {
	initialized bool
	mutex       sync.Mutex
}

var bootstrap = &YcsBootstrap{}

// Initialize initializes the Ycs system with dependency injection to avoid circular references
func Initialize() {
	bootstrap.mutex.Lock()
	defer bootstrap.mutex.Unlock()

	if bootstrap.initialized {
		return
	}

	// Register type readers directly in ContentType to match C# implementation
	content.RegisterTypeReader(0, func(decoder contracts.IUpdateDecoder) contracts.IAbstractType {
		return core.ReadYArray(decoder)
	}) // YArray
	content.RegisterTypeReader(1, func(decoder contracts.IUpdateDecoder) contracts.IAbstractType {
		return core.ReadYMap(decoder)
	}) // YMap
	content.RegisterTypeReader(2, func(decoder contracts.IUpdateDecoder) contracts.IAbstractType {
		return core.ReadYText(decoder)
	}) // YText

	// Set document factory
	content.SetDocFactory(func(opts *contracts.YDocOptions) contracts.IYDoc {
		return core.NewYDoc(*opts)
	})

	bootstrap.initialized = true
}

// IsInitialized returns whether the bootstrap has been initialized
func IsInitialized() bool {
	bootstrap.mutex.Lock()
	defer bootstrap.mutex.Unlock()
	return bootstrap.initialized
}

File: src/Ycs/YcsBootstrap.cs
// ------------------------------------------------------------------------------
//  <copyright company="Microsoft Corporation">
//      Copyright (c) Microsoft Corporation.  All rights reserved.
//  </copyright>
// ------------------------------------------------------------------------------

using Ycs.Content;
using Ycs.Contracts;
using Ycs.Core;

namespace Ycs
{
    /// <summary>
    /// Bootstrap class to initialize dependencies and break circular references
    /// </summary>
    public static class YcsBootstrap
    {
        private static bool _initialized = false;

        
        /// <summary>
        /// Initialize the Ycs system with dependency injection to avoid circular references
        /// </summary>
        public static void Initialize()
        {
            if (_initialized)
            {
                return;
            }

            ContentType.RegisterTypeReader(0, decoder => YArray.Read(decoder)); // YArray
            ContentType.RegisterTypeReader(1, decoder => YMap.Read(decoder));   // YMap  
            ContentType.RegisterTypeReader(2, decoder => YText.Read(decoder));  // YText

            ContentDoc.SetDocFactory(opts => new YDoc(opts));
            _initialized = true;
        }
    }
}